onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.cc

```diff
@@ -372,6 +388,113 @@ Status ProcessLogits(const OrtValue& logits,                                 //
+
+  const T* logits_data = logits.Get<Tensor>().Data<T>();
+
+  // Logits has shape (batch_size, input_length, vocab_size),
+  // where input_length equals to parameters_->sequence_length for first subgraph call, and 1 for the remaining calls.
+  const TensorShape& logits_shape = logits.Get<Tensor>().Shape();
+  ORT_ENFORCE(logits_shape.NumDimensions() == 3);
+  auto input_length = logits_shape[1];
+
+  // Get logits for the last token:
+  //    next_token_logits = logits[:, -1, :], and the result shape is (batch_size, vocab_size)
+  // When input_length == 1, use logits directly in SoftmaxCPU below so it only need for input_length > 1.
+  gsl::span<T>& next_token_scores = greedy_state->next_token_scores;
+  const T* current_logits = logits_data + (input_length - 1) * vocab_size;
+  for (int i = 0; i < batch_size; i++) {
+    gsl::span<const T> source(current_logits, vocab_size);
+    gsl::span<T> target = next_token_scores.subspan(SafeInt<gsl::index>(i) * vocab_size,
+                                                    static_cast<gsl::index>(vocab_size));
+    gsl::copy(source, target);
+    current_logits += input_length * vocab_size;
+  }
+
+#ifdef DEBUG_GENERATION
+  dumper->Print("logits", logits);
+  dumper->Print("next_token_logits", next_token_scores.data(), batch_size, 1, vocab_size);
+#endif
+
+  // Apply all score processors that updates scores
+  logits_processors->Process(sequences, next_token_scores, step);
+
+#ifdef DEBUG_GENERATION
+  dumper->Print("next_token_scores after logits processor", next_token_scores.data(), batch_size, 1, vocab_size);
+#endif
+
+  // TODO(wy): support output_scores
+  ORT_UNUSED_PARAMETER(output_scores);
+
+  // next_tokens = torch.argmax(scores, dim=-1)
+  int64_t next_token_scores_dims[] = {static_cast<int64_t>(batch_size), vocab_size};
+  TensorShape next_token_scores_shape(&next_token_scores_dims[0], 2);
+  auto element_type = DataTypeImpl::GetType<T>();
+  OrtValue next_token_scores_value;
+  Tensor::InitOrtValue(element_type,
+                       next_token_scores_shape,
+                       next_token_scores.data(),
+                       allocator->Info(),
+                       next_token_scores_value);
+  const Tensor& input = next_token_scores_value.Get<Tensor>();
+
+  constexpr int axis = 1;
+  const unsigned top_k = static_cast<unsigned>(1);
+  constexpr bool largest = true;
+  constexpr bool sorted = false;
+
+  std::unique_ptr<Tensor> topk_scores;
+  std::unique_ptr<Tensor> topk_indices;
+  ORT_RETURN_IF_ERROR(
+    TopK(&input,
+         axis,
+         top_k,
+         largest,
+         sorted,
+         allocator,
+         stream,
+         thread_pool,
+         topk_scores,
+         topk_indices));
+
+#ifdef DEBUG_GENERATION
+  dumper->Print("topk_scores", *(topk_scores.get()));
+  dumper->Print("topk_indices", *(topk_indices.get()));
+#endif
+
+  gsl::span<const int64_t> next_token_indices = topk_indices->DataAsSpan<int64_t>();
+  for (int i = 0; i < batch_size; i++) {
```

Comment:
why the logic in beam search not applied here? (https://github.com/microsoft/onnxruntime/blob/52095fb042ebcb9f1e520d58b23fb9cec260aab3/onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc#L347-L354)?